File: USACO_Problems/cycle_correspondence/cyc_corr.cpp
#include <iostream>
#include <vector>
#include <algorithm> // for std::find and std::distance

int main() {
    int n, k;
    std::cin >> n >> k;
    std::vector<int> list_a(k), list_b(k);
    for (int i = 0; i < k; ++i) std::cin >> list_a[i];
    for (int i = 0; i < k; ++i) std::cin >> list_b[i];

    std::vector<bool> included(n + 1, false);
    int not_inc = 0;

    for (int i = 0; i < k; ++i) {
        included[list_a[i]] = true;
        included[list_b[i]] = true;
    }

    for (int i = 1; i <= n; ++i) {
        if (!included[i]) not_inc++;
    }

    std::vector<int> offset;
    for (int i = 0; i < k; ++i) {
        auto it = std::find(list_b.begin(), list_b.end(), list_a[i]);
        if (it != list_b.end()) {
            int index = std::distance(list_b.begin(), it);
            if (index - i < 0) offset.push_back(index - i + k);
            else offset.push_back(index - i);
        }
    }

    int length = offset.size();
    std::vector<int> offset_counts;
    int count = 0;
    if (length != 0) {
        for (int i = 0; i < length - 1; ++i) {
            if (offset[i] == offset[i + 1]) count++;
            else {
                count++;
                offset_counts.push_back(count);
                count = 0;
            }
        }
        count++;
        offset_counts.push_back(count);
    }
    int answer = length == 0 ? 0 : not_inc + *std::max_element(offset_counts.begin(), offset_counts.end());

    std::reverse(list_a.begin(), list_a.end());
    offset.clear();

    for (int i = 0; i < k; ++i) {
        auto it = std::find(list_b.begin(), list_b.end(), list_a[i]);
        if (it != list_b.end()) {
            int index = std::distance(list_b.begin(), it);
            if (index - i < 0) offset.push_back(index - i + k);
            else offset.push_back(index - i);
        }
    }

    std::sort(offset.begin(), offset.end());
    length = offset.size();
    offset_counts.clear();
    count = 0;
    if (length != 0) {
        for (int i = 0; i < length - 1; ++i) {
            if (offset[i] == offset[i + 1]) count++;
            else {
                count++;
                offset_counts.push_back(count);
                count = 0;
            }
        }
        count++;
        offset_counts.push_back(count);
    }
    answer = length == 0 ? 0 : std::max(answer, not_inc + *std::max_element(offset_counts.begin(), offset_counts.end()));

    std::cout << answer << std::endl;

    return 0;
}

File: USACO_Problems/cycle_correspondence/cyc_corr.py
n, k = list(map(int, input().split()))
list_a = list(map(int, input().split()))
list_b = list(map(int, input().split()))

offset = []

included = [False] * (1 + n) # could be made better
not_inc = 0

for i in range(k): # could be made better
    included[list_a[i]] = True 
    included[list_b[i]] = True
    
for i in range(1, 1 + n): # could be made better
    if not included[i]:
        not_inc += 1

for i in range(k):
    if list_a[i] in list_b: # append 0?
        if list_b.index(list_a[i]) - i < 0:
            offset.append(list_b.index(list_a[i]) - i + k)
        else:
            offset.append(list_b.index(list_a[i]) - i)

length = len(offset)
offset_counts = []
count = 0        
if length != 0:
    for i in range(length - 1):
        if offset[i] == offset[i + 1]:
            count += 1
        else:
            count += 1
            offset_counts.append(count)
            count = 0
    count += 1
    offset_counts.append(count)
    count = 0
    answer = not_inc + max(offset_counts)
else:
    answer = 0


list_a = list(reversed(list_a))
offset = [] 

for i in range(k):
    if list_a[i] in list_b:
        if list_b.index(list_a[i]) - i < 0:
            offset.append(list_b.index(list_a[i]) - i + k)
        else:
            offset.append(list_b.index(list_a[i]) - i)
        
        
offset.sort()
length = len(offset)
offset_counts = []
count = 0
if length != 0:
    for i in range(length - 1):
        if offset[i] == offset[i + 1]:
            count += 1
        else:
            count += 1
            offset_counts.append(count)
            count = 0
    count += 1
    offset_counts.append(count)
    count = 0
    answer = max(answer, not_inc + max(offset_counts))
else:
    answer = 0
    
print(answer)

# n, k = list(map(int, input().split()))
# list_a = list(map(int, input().split()))
# list_b = list(map(int, input().split()))

# offset = []

# included = [False] * (1 + n)
# answer = 0

# for i in list_a:
#     included[i] = True
    
# for i in list_b:
#     included[i] = True

# for i in range(k):
#     if list_a[i] in list_b:
#         if list_b.index(list_a[i]) - i < 0:
#             offset.append(list_b.index(list_a[i]) - i + k)
#         else:
#             offset.append(list_b.index(list_a[i]) - i)
#--------------------            
# offset_counts = []
# count = 0        
# if len(offset) != 0:
#     for i in range(len(offset) - 1):
#         if offset[i] == offset[i + 1]:
#             count += 1
#         else:
#             count += 1
#             offset_counts.append(count)
#             count = 0
#     count += 1
#     offset_counts.append(count)
#     count = 0
#-------------
# offset = []            
# list_a = list(reversed(list_a))
# for i in range(k):
#     if list_a[i] in list_b:
#         if list_b.index(list_a[i]) - i < 0:
#             offset.append(list_b.index(list_a[i]) - i + k)
#         else:
#             offset.append(list_b.index(list_a[i]) - i)
        
# for i in range(1, 1 + n):
#     if not included[i]:
#         answer += 1
# offset.sort()

# offset_counts = []
# count = 0
# if len(offset) != 0:
#     for i in range(len(offset) - 1):
#         if offset[i] == offset[i + 1]:
#             count += 1
#         else:
#             count += 1
#             offset_counts.append(count)
#             count = 0
#     count += 1
#     offset_counts.append(count)
#     count = 0
#     print(answer + max(offset_counts))
# else:
#     print(0)